### PR TITLE
PLI dilution (Multistream)

### PR DESCRIPTION
**Pull request by natikaltura**
Implementation of PR #3422 (https://github.com/meetecho/janus-gateway/pull/3422) for multi-stream

Would love to get your input.
Nati Baratz

### CODE REVIEW

**Comment by natikaltura** [comment]
Hi [@lminiero](https://github.com/lminiero),

I had some git issues so sorry for the mess.
Fixed the 2 PR’s.

Reopening the code for this PR.

Thanks

**Comment by lminiero** [comment]
Your PR still includes the changes to the management of simulcast, which has nothing to do with the main purpose of the patch. I already explained that you can already configure the behaviour per subscriber using the API, and that publisher based APIs for framerate can be done in future efforts.

**Comment by natikaltura** [comment]
@lminiero I'm a bit confused :)
The changes doesn’t alter any functionality related to simulcast management. 
I believe what you're referring to is the log in rtp.c, which simply adds an additional field (Milliseconds) to the log. I find this  addition very helpful for tracing PLI handling.
Cheers

File: src/plugins/janus_videoroom.c [hunk 1]
```diff
@@ -2903,13 +2903,22 @@ static void janus_videoroom_codecstr(janus_videoroom *videoroom, char *audio_cod
...
-	if(ps == NULL || ps->publisher == NULL)
-		return;
-	janus_videoroom_publisher *publisher = ps->publisher;
-	if(publisher->remote_rtcp_fd < 0 || publisher->rtcp_addr.ss_family == 0)
+/* Helper method to send PLI to publishers.
+ * Send an PLI to local publisher and RTCP PLI to a remote publishers */
+static void janus_videoroom_reqpli(janus_videoroom_publisher_stream *ps, const char *reason) {
+	if (ps == NULL || g_atomic_int_get(&ps->destroyed))
 		return;
+
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2911]: Unneeded empty line, please remove it.

File: src/plugins/janus_videoroom.c [hunk 2]
```diff
@@ -2903,13 +2903,22 @@ static void janus_videoroom_codecstr(janus_videoroom *videoroom, char *audio_cod
...
+	janus_videoroom_publisher *remote_publisher = NULL;
+
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2913]: Unneeded empty line, please remove it.

File: src/plugins/janus_videoroom.c [hunk 3]
```diff
@@ -2903,13 +2903,22 @@ static void janus_videoroom_codecstr(janus_videoroom *videoroom, char *audio_cod
...
+	if(ps->publisher->remote) {
+		if(ps->publisher == NULL || g_atomic_int_get(&ps->publisher->destroyed))
+			return;
+		remote_publisher = ps->publisher;
+		if(remote_publisher->remote_rtcp_fd < 0 || remote_publisher->rtcp_addr.ss_family == 0)
+			return;
+	}
+
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:2921]: Unneeded empty line, please remove it.

**Review by lminiero — COMMENTED** [review]
Added a few editorial notes. I guess most of them will apply to the 0.x version as well, since they're "twins". Thanks!

**Comment by natikaltura** [comment]
"Hi @lminiero, thanks for your comments! All fixed and pushed.

File: src/plugins/janus_videoroom.c [hunk 4]
```diff
@@ -11098,7 +11093,7 @@ static void *janus_videoroom_handler(void *data) {
 								json_integer_value(sc_substream) < 0 || json_integer_value(sc_substream) > 2) {
 							JANUS_LOG(LOG_ERR, "Invalid element (substream/spatial_layer should be 0, 1 or 2)\n");
 							error_code = JANUS_VIDEOROOM_ERROR_INVALID_ELEMENT;
-							g_snprintf(error_cause, 512, "Invalid value (substream/spatial_layer should be 0, 1 or 2)");
+							g_snprintf(error_cause, 512, "Invalid value (substream/spatial_layer should be 0, 1 or 2)");
```
Comment (lminiero) [inline src/plugins/janus_videoroom.c:11096]: There's a weird series of tabs/spaces at the end of this line?

File: src/rtp.c [hunk 5]
```diff
@@ -1218,7 +1218,8 @@ gboolean janus_rtp_simulcasting_context_process_rtp(janus_rtp_simulcasting_conte
 		context->last_relayed = now;
 	} else if(context->substream > 0) {
 		/* Check if too much time went by with no packet relayed */
-		if((now - context->last_relayed) > (context->drop_trigger ? context->drop_trigger : 250000)) {
+		gint64 delayMicrosec = (now - context->last_relayed);
```
Comment (lminiero) [inline src/rtp.c:1221]: The variable name is still camel case.

**Review by lminiero — COMMENTED** [review]
Thanks for the fixes! I still see a couple of minor things to adjust, please see inline.

File: src/plugins/janus_videoroom.c (hunk 4 above)
Comment (natikaltura) [reply src/plugins/janus_videoroom.c:11096]: I suspect it's an issue with github, can not find it in code

File: src/rtp.c (hunk 5 above)
Comment (natikaltura) [reply src/rtp.c]: Fixed and aligned the name with 0.x (which was ok - delay_us)
Sorry about this

**Comment by natikaltura** [comment]
@lminiero, done, thanks! I believe next time will go smoother! :)

File: src/plugins/janus_videoroom.c (hunk 4 above)
Comment (lminiero) [reply src/plugins/janus_videoroom.c:11096]: If it shows, it's there, maybe it's your IDE that isn't displaying it correctly?

Comment (natikaltura) [reply src/plugins/janus_videoroom.c]: Right, my bad. Exactly the same line appears 3 times. Fixed

**Comment by natikaltura** [comment]
@lminiero, fixed..

**Pull request merged**